src/ucs/memory/rcache.c

```diff
@@ -1060,6 +1075,15 @@ static ucs_status_t ucs_rcache_global_list_add(ucs_rcache_t *rcache)
 static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache) {
     pthread_mutex_lock(&ucs_rcache_global_context.lock);
     ucs_list_del(&rcache->list);
+    if (!ucs_list_is_empty(&ucs_rcache_global_context.list)) {
+        goto out;
+    }
+
+    ucs_async_remove_handler(ucs_async_pipe_rfd(
```

Comment:
Calling `ucs_async_remove_handler` without holding the lock results in an "invalid fd" error.
Maybe calling it with the lock and without `is_sync=1` would be a good idea?

Reply:
calling with is_sync=0 can lead to handler being called after fd is destroyed
what is the reason it can result in invalid fd? can you post the backtrace of it?

Reply:
@yosefe - had a race condition. Fixed.